acados/ocp_nlp/ocp_nlp_sqp.c

```diff
@@ -785,76 +701,271 @@ int ocp_nlp_sqp(void *config_, void *dims_, void *nlp_in_, void *nlp_out_,
+
+                // // tmp_nlp_out = iterate + step
+                // for (ii = 0; ii <= N; ii++)
+                //     blasfeo_daxpy(nv[ii], 1.0, qp_out->ux+ii, 0, nlp_out->ux+ii, 0, work->nlp_work->tmp_nlp_out->ux+ii, 0);
+
+    //             // evaluate
+    // #if defined(ACADOS_WITH_OPENMP)
+    //     #pragma omp parallel for
+    // #endif
+    //             for (ii=0; ii<N; ii++)
+    //             {
+    //                 config->dynamics[ii]->compute_fun(config->dynamics[ii], dims->dynamics[ii], nlp_in->dynamics[ii],
+    //                                                 nlp_opts->dynamics[ii], nlp_mem->dynamics[ii], work->nlp_work->dynamics[ii]);
+    //             }
+    // #if defined(ACADOS_WITH_OPENMP)
+    //     #pragma omp parallel for
+    // #endif
+    //             for (ii=0; ii<=N; ii++)
+    //             {
+    //                 config->constraints[ii]->compute_fun(config->constraints[ii], dims->constraints[ii],
+    //                                                     nlp_in->constraints[ii], nlp_opts->constraints[ii],
+    //                                                     nlp_mem->constraints[ii], work->nlp_work->constraints[ii]);
+    //             }
+    // #if defined(ACADOS_WITH_OPENMP)
+    //     #pragma omp parallel for
+    // #endif
+                // update QP rhs
+                // d_i = c_i(x_k + p_k) - \nabla c_i(x_k)^T * p_k
+                struct blasfeo_dvec *tmp_fun_vec;
+
+                for (ii = 0; ii <= N; ii++)
+                {
+                    if (ii < N)
+                    {
+                        // b -- dynamics
+                        tmp_fun_vec = config->dynamics[ii]->memory_get_fun_ptr(nlp_mem->dynamics[ii]);
+                        // add - \nabla c_i(x_k)^T * p_k
+                        blasfeo_dgemv_t(nx[ii]+nu[ii], nx[ii], -1.0, qp_in->BAbt+ii, 0, 0,
+                                        qp_out->ux+ii, 0, 1.0, tmp_fun_vec, 0, qp_in->b+ii, 0);
+                        blasfeo_dvecad(nx[ii+1], 1.0, qp_out->ux+ii+1, nu[ii+1], qp_in->b+ii, 0);
+                    }
+
+                    /* INEQUALITIES */
+                    // d -- constraints
+                    tmp_fun_vec = config->constraints[ii]->memory_get_fun_ptr(nlp_mem->constraints[ii]);
+                    /* SOC for bounds can be skipped (because linear) */
+                    // NOTE: SOC can also be skipped for truely linear constraint, i.e. ng of nlp, now using ng of QP = (nh+ng)
+
+                    // upper & lower
+                    blasfeo_dveccp(ng[ii], tmp_fun_vec, nb[ii], qp_in->d+ii, nb[ii]); // lg
+                    blasfeo_dveccp(ng[ii], tmp_fun_vec, 2*nb[ii]+ng[ii], qp_in->d+ii, 2*nb[ii]+ng[ii]); // ug
+                    // general linear / linearized!
+                    // d[nb:nb+ng] += D * u + C * x (lower)
+                    blasfeo_dgemv_t(nu[ii]+nx[ii], ng[ii], 1.0, qp_in->DCt+ii, 0, 0, qp_out->ux+ii, 0,
```

Comment:
Since the constraints are two sided, you only need to perform the matrix-vector multiplication (the most expensive operation) once and cache it in a working vector, to reuse for both upper and lower constraint.

Reply:
Right!
Addressed in: https://github.com/acados/acados/pull/772/commits/39e7718c14bcbb204ae0ae0b6e9c5c1fd799fbc7